actor/kvcounter-wasmcon2023/src/lib.rs

```diff
@@ -0,0 +1,235 @@
+            // If the value exists, parse it into an i32
+            Ok(incoming_value) => {
+                // Read bytes from incoming value
+                let bytes = incoming_value_consume_sync(incoming_value)
+                    .map_err(|count| anyhow!("failed to parse incoming bytes, read [{count}]"))?;
+                // Convert the bytes to a i32
+                String::from_utf8(bytes)
+                    .context("failed to parse string from returned bytes")?
+                    .trim()
+                    .parse()
+                    .context("failed to parse i32 from bytes")?
+            }
+            // If the value is missing or we fail to get it, assume it is zero
+            Err(_) => {
+                eprintln!("[warn] encountered missing key [{key}], defaulting to 0");
+                0
+            }
+        };
+
+        // Calculate the new value
+        let new_value: i32 = current_value + amount;
+
+        // Build outgoing value to use
+        let outgoing_value = new_outgoing_value();
+        let stream =
+            outgoing_value_write_body(outgoing_value).expect("failed to write outgoing value");
+
+        // Write out the new value
+        write(stream, new_value.to_string().as_bytes())
+            .expect("failed to write to outgoing value stream");
+
+        // Set the key to the updated value
+        set(bucket, key, outgoing_value).expect("failed to set value");
+
+        // Cheat and just assume the new value will be the increment
+        Ok(new_value)
+    }
+}
+
+/// Write a W HTTP response out, using WIT-driven (WASI) interfaces
+fn write_http_response(
+    response_outparam: ResponseOutparam,
+    status_code: u16,
+    headers: &[(String, Vec<u8>)],
+    body: impl AsRef<[u8]>,
+) {
+    // Add headers
+    let headers = new_fields(headers);
+
+    // Create new outgoing response and related stream
+    let outgoing_response =
+        new_outgoing_response(status_code, headers).expect("failed to create response");
+    let outgoing_stream =
+        outgoing_response_write(outgoing_response).expect("failed to write outgoing response");
+
+    // Write out repsonse body to outgoing straem
+    write(outgoing_stream, body.as_ref()).expect("failed to write output to stream");
+    finish_outgoing_stream(outgoing_stream);
+
+    // Set the response on the param
+    set_response_outparam(response_outparam, Ok(outgoing_response))
+        .expect("failed to set response");
+}
+
+/// Implementation of the WIT-driven incoming-handler interface for our implementation struct
+impl IncomingHandler for KvCounter {
+    fn handle(request: IncomingRequest, response: ResponseOutparam) {
+        // Decipher method
+        let method = incoming_request_method(request);
+
+        // Decipher path
+        let path_with_query = incoming_request_path_with_query(request).expect("invalid path");
+        let path = match path_with_query.split_once('?') {
```

Comment:
Heads up: I was doing this with blobby but I have the `http` crate imported and it has a helpful parser for it: `http::uri::PathAndQuery`

Reply:
Ahh thank you that's really helpful, I'll use that!

- [ ] Use `http::uri::PathAndQuery` for parsing incoming request path